HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,390 @@
+
+  // Number of objects passing the L3 Trigger or L2:
+  int n = 0;
+  bool firstIteration = true;
+  for (unsigned int iMuL2 = 0; iMuL2 < L2mucands.size(); iMuL2++) {
+    reco::RecoChargedCandidateRef L2cand(iEvent.getHandle(L2CandToken_), iMuL2);
+    LogDebug("HLTMuonL3orL2PreFilter") << "cand isNonnull " << L2cand.isNonnull();
+
+    //    //did this candidate triggered at previous stage.
+    //    if (matchPreviousCand_ && !triggerdByPreviousLevel(cand, vcands))
+    //      continue;
+    bool FailedL2 = false;
+    if (std::abs(L2cand->eta()) > max_Eta_L2_)
+      FailedL2 = true;
+
+    reco::TrackRef L2tk = L2cand->track();
+    LogDebug("HLTMuonL3orL2PreFilter") << " Muon in loop, q*pt= " << L2tk->charge() * L2tk->pt() << " ("
+                                       << L2cand->charge() * L2cand->pt() << ") "
+                                       << ", eta= " << L2tk->eta() << " (" << L2cand->eta() << ") "
+                                       << ", hits= " << L2tk->numberOfValidHits() << ", d0= " << L2tk->d0()
+                                       << ", dz= " << L2tk->dz();
+
+    //normalizedChi2 cut
+    if (L2tk->normalizedChi2() > max_NormalizedChi2_L2_)
+      FailedL2 = true;
+
+    if (not recoBeamSpotHandle.isValid()) {
+      throw cms::Exception("MissingProduct") << "can not find " << beamspotTag_;
+    }  //Throw an exception if the beamSpot is not valid
+
+    if (recoBeamSpotHandle.isValid()) {
+      const reco::BeamSpot& beamSpot = *recoBeamSpotHandle;
+
+      //dz cut (impact parameter of the muon in the z direction computed with respect to the beamspot position)
+      float vz_L2 = L2cand->vz() - beamSpot.z0();
+      float vx_L2 = L2cand->vx() - beamSpot.x0();
+      float vy_L2 = L2cand->vy() - beamSpot.y0();
+      if (std::abs(vz_L2 - (vx_L2 * L2cand->px() + vy_L2 * L2cand->py()) / (L2cand->pt() * L2cand->pt()) *
+                               L2cand->pz()) > max_Dz_)
+        FailedL2 = true;
+
+      // dxy significance cut (safeguard against bizarre values)
+      float const absDxy = std::abs(L2tk->dxy(beamSpot.position()));
+      if (min_DxySig_ > 0 && (L2tk->dxyError() <= 0 || (absDxy < min_DxySig_ * L2tk->dxyError())))
+        FailedL2 = true;
+
+      //dxy beamspot cut
+      if (absDxy > max_DXYBeamSpot_ || absDxy < min_DXYBeamSpot_)
+        FailedL2 = true;
+    }
+
+    //min muon hits cut
+    const reco::HitPattern& L2trackHits = L2tk->hitPattern();
+    if (L2trackHits.numberOfValidMuonHits() < min_NmuonHits_L2_)
+      FailedL2 = true;
+
+    // number of eta bins for cut on number of stations
+    const std::vector<double>::size_type nAbsetaBins = absetaBins_L2_.size();
+
+    // cut on number of stations
+    bool failNstations(false), failNhits(false), failNchambers(false);
+    for (unsigned int i = 0; i < nAbsetaBins; ++i) {
+      if (std::abs(L2cand->eta()) < absetaBins_L2_[i]) {
+        if (L2cand->track()->hitPattern().muonStationsWithAnyHits() < minNstations_L2_[i]) {
```

Comment:
I don't understand why do we have to check this. What is the reason?

Reply:
The reason is fairly obvious. If the PSet is misconfigured, for example with `absetaBins_L2_` having a larger size than `minNstations_L2_`, then `minNstations_L2_[i]` in L237 will cause a segmentation violation. The same reasoning applies to the other relevant vectors. In general, the logic of this module demands these vectors have the same size, and that's why this needs to be checked explicitly in the ctor; in case those sizes differ, one can get a seg-fault (see above), or possibly wrong physics outputs in the worst case scenario.

Reply:
Just noting that this point has not been addressed yet, as far as I can see. It is a simple update, and one can follow the example linked above.

Reply:
@alexsr-98 

Since this point is not optional, I'm waiting for this to be addressed before proceeding further with the review.